Improve MAX! Cube integration

leppa opened the pull request:
## Breaking Change:

Custom `vacation` preset was replaced with standard `away`. Additionally, `manual` preset was removed in favor of HVAC mode `heat`. This might affect people's scripts, templates or automations.

## Description:

A few improvements to the MAX! Cube platform:

1. It wasn't possible to return to automatic (scheduled) operation after switching to `manual` preset. `none` preset was added, which switches thermostats to automatic (scheduled) operation.

2. *Manual* preset was removed. HVAC mode *Heat* now switches thermostats into manual mode.

3. `hvac_action` was implemented, based on the valve position: if it's open (position higher than 0), we assume that hot water is flowing and heating is active. Else, the thermostat is in *Idle* mode. For wall thermostats, if at least one thermostat in the room has its valve open, then heating is active.

4. Support for *Comfort*, *Eco*, *Off* and *On* modes was added. *Comfort*, *Eco* and *On* are exposed as presets and switch thermostats into manual mode with corresponding temperatures, programmed on the thermostats. *Off* corresponds to HVAC mode *Off*.

5. Valve position is exposed as state attribute.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.
  - [x] I have followed the [development checklist][dev-checklist]

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html

leppa committed:
maxcube: Make it possible to return to auto mode

It's currently not possible to switch back to auto (schedule) mode after
changing to manual or vacation. Add `PRESET_NONE` to list of presets
that will return thermostat back into auto (schedule) mode.

Additionally, use `PRESET_AWAY` instead of custom preset name for
vacation mode.

leppa committed:
maxcube: Provide `hvac_action` based on valve state

Not supported for wall thermostats.

leppa committed:
maxcube: Add support for Comfort, Eco, Off and On modes

Off is mapped to HVAC_OFF, while On - to HVAC_HEAT.

leppa committed:
maxcube: Add `hvac_action` support for wall thermostats

We check all thermostats in the same room as the wall thermostat. If at
least one of them has its valve open - the room is being heated.

leppa committed:
maxcube: Expose valve position as state attribute

Also fix a small logical error in `hvac_action`.

leppa committed:
maxcube: Fix linter errors and formatting

ishkliar commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -95,13 +109,73 @@ def current_temperature(self):
+        temp = device.target_temperature
+        mode = device.mode
+
+        if hvac_mode == HVAC_MODE_OFF:
```
I think we shouldn't use these tricks if the thermostat doesn't support heat/off mode

ishkliar replied:
I think that HVAC_MODE_AUTO and HVAC_MODE_HEAT are enough.

leppa replied:
Device claims to support On (valves fully open - always heat) and Off modes (valves fully closed - never heat), but they're implemented weirdly, through the magic temperatures. When you set the thermostat to Off (through the dial on the thermostat), the display will show Off, but the reported target temperature through the API will be 4.5 degrees. Same with On, but temperature is 30.5 degrees.

ishkliar replied:
So, I've tested this solution in my home environment and I like it.

leppa committed:
maxcube: Adjust mapping between MAX! and HA modes and presets

MAX! 'Manual' mode now corresponds to 'Heating' mode of HA. MAX! 'Eco'
and 'Comfort' temperatures are 'Heating' mode presets. MAX! 'On' mode is
now a 'Heating' preset, too.

leppa commented:
After paying attention to a commend about `HVAC_MODE_AUTO` in `const.py` of the climate component, it looks like `HVAC_MODE_AUTO` actually means that *target temperature* is controlled automatically, and not that the device automatically keeps the room temperature according to the target temperature. So I have adjusted the mapping between MAX! Cube and HA modes / presets. The `HVAC_MODE_AUTO` now only corresponds to scheduled (`PRESET_NONE`), *Boost* (`PRESET_BOOST`) and *Vacation* (`PRESET_AWAY`) modes, as those will adjust the target temperature based on schedule or timeout. *Eco* (`PRESET_ECO`), *Comfort* (`PRESET_COMFORT`) and *On* (custom `PRESET_ON`) modes are presets of the `HVAC_MODE_MANUAL`, as those keep target temperature constant. *Off* still corresponds to the `HVAC_MODE_OFF`.

dYalib commented:
Hi, @leppa 
unfortunately i saw your work to late, so i have do a pull request https://github.com/home-assistant/home-assistant/pull/30420 with the same goal as you. I think it's a good idea when we merge our work together. I will test your solution and give feedback.

leppa commented:
> Hi, @leppa
> unfortunately i saw your work to late, so i have do a pull request #30420 with the same goal as you. I think it's a good idea when we merge our work together. I will test your solution and give feedback.

Hi @dYalib,

I quickly checked your change and it looks like my change more or less incorporates your. Feel free to test it and provide feedback.

frenck commented:
Thanks, stale bot, however, this integration did not have an initial review yet. Please unmark it. Thanks.

MartinHjelmare reviewed (changes requested):
Sorry for the late review. Looks good! Some smaller comments and questions.

leppa commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -95,13 +109,72 @@ def current_temperature(self):
     def hvac_mode(self):
-        """Return current operation (auto, manual, boost, vacation)."""
+        """Return current operation mode."""
+        device = self._cubehandle.cube.device_by_rf(self._rf_address)
```
As far as I understand, yes. You need to call `self._cubehandle.update()` to refresh the values from MAX! Cube.

leppa commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -95,13 +109,72 @@ def current_temperature(self):
+            if device.target_temperature == OFF_TEMPERATURE:
+                return HVAC_MODE_OFF
+            return HVAC_MODE_HEAT
         return HVAC_MODE_AUTO
```
The behavior of MAX! Cube in automatic / scheduled mode (`MAX_DEVICE_MODE_AUTOMATIC`) is that it still allows user to change target temperature, but it will be temporary. The temperature will be reset to scheduled one as soon as next setpoint time comes or if you open and close the window. To preserve user set temperature, MAX! Cube has to be set into manual operation mode, which I mapped to Home Assistant's `HVAC_MODE_HEAT`. If I'm not mistaken, vacation mode (`MAX_DEVICE_MODE_VACATION`) behaves the same as automatic mode. However, there is only one setpoint time - it exits vacation mode and switches to automatic one.

MartinHjelmare replied:
Eg Google assistant will not allow changing the temperature in our `HVAC_MODE_AUTO`. I'm not sure if there's a better mapping. For greater flexibility in the modes we recommend using the preset feature since the integration can define presets itself. Maybe extend the presets instead?

leppa replied:
I don't think that MAX! Cube's automatic mode intended temperature customization. Because any changes will be reset by the Cube as soon as some event happens (like schedule-relevant time change, window open / close). It's just that they still allow you to do that. For me this behaviour is much closer to `HVAC_MODE_AUTO` than to `HVAC_MODE_HEAT`. If someone wants to change temperature to something different from currently scheduled one, they would most certainly want to preserve it, so they would have to change to manual mode anyway.

So I would strongly prefer leaving mapping of `MAX_DEVICE_MODE_AUTOMATIC` to `HVAC_MODE_AUTO`.

Vacation mode is intended to be used when you leave for longer periods of time, so that heating returns to automatic mode at some predefined date (e.g., the date when you return back). There's no-one at home to change the temperature while you're on vacation.

`MAX_DEVICE_MODE_VACATION` is already mapped to `PRESET_AWAY`. The only question is which HVAC mode sounds more logical for this preset - `HVAC_MODE_HEAT` or `HVAC_MODE_AUTO`?

MartinHjelmare replied:
Can the user set the temperature in vacation mode?

leppa replied:
You can select temperature when you activate vacation mode. You should also set a date when vacation mode should end (which doesn't seem to be supported by HA). I just tried activating vacation mode and changing temperature directly on the thermostat - it doesn't let me. It even ignores opening / closing of windows. However, I can change temperature through their official MAX! App. Again, by the behavior of the thermostat, it looks like it's not supposed to be changed, but API still allows it.

MartinHjelmare replied:
If the temperature will be fixed during the vacation mode, it sounds best to set the home assistant mode to `HVAC_MODE_HEAT` when selecting the away preset.

leppa replied:
Yeah, it's supposed to be fixed up until vacation mode expires. So it's `HVAC_MODE_HEAT`, then.

MartinHjelmare replied:
Good!

leppa commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -95,13 +109,72 @@ def current_temperature(self):
+                cube.set_temperature_mode(device, temp, mode)
+            except (socket.timeout, socket.error):
+                _LOGGER.error("Setting HVAC mode failed")
+                return False
```
Not sure why `False` got there. Most probably, a result of copy-paste. I'll log error and simply `return`.

leppa commented:
@MartinHjelmare, thanks for the review. I answered your questions and will try to address remaining comments during this week.

leppa committed:
maxcube: Address review comments

MartinHjelmare commented on homeassistant/components/maxcube/climate.py:
```diff
@@ -42,6 +42,20 @@
+    PRESET_AWAY: MAX_DEVICE_MODE_VACATION,
+    PRESET_BOOST: MAX_DEVICE_MODE_BOOST,
+    PRESET_NONE: MAX_DEVICE_MODE_AUTOMATIC,
+    PRESET_ON: MAX_DEVICE_MODE_MANUAL,
```
Should it be `PRESET_ON` or `PRESET_NONE`?

MartinHjelmare replied:
What should it be then? I'm asking since this isn't mapped the same way in the other dict below.

leppa replied:
The problem is that there's no one-to-one mapping between MAX! Cube modes and HA HVAC modes. Proper mapping depends on additional data. Here's a table of how it looks:

| MAX! Cube UI | MAX! Cube Mode | MAX! Cube target t°C | HA HVAC mode | HA preset |
| --- | --- | --- | --- | --- |
| Auto (schedule) | `MAX_DEVICE_MODE_AUTOMATIC` | *any* | `HVAC_MODE_AUTO` | `PRESET_NONE` |
| Boost | `MAX_DEVICE_MODE_BOOST` | *any* | `HVAC_MODE_AUTO` | `PRESET_BOOST` |
| Vacation | `MAX_DEVICE_MODE_VACATION` | *any* | `HVAC_MODE_MANUAL` | `PRESET_AWAY` |
| Eco | `MAX_DEVICE_MODE_MANUAL`¹ | User configured Eco² | `HVAC_MODE_MANUAL` | `PRESET_ECO` |
| Comfort | `MAX_DEVICE_MODE_MANUAL`¹ | User configured Comfort² | `HVAC_MODE_MANUAL` | `PRESET_COMFORT` |
| On | `MAX_DEVICE_MODE_MANUAL`¹ | 30.5 | `HVAC_MODE_MANUAL` | `PRESET_ON` |
| Off | `MAX_DEVICE_MODE_MANUAL`¹ | 4.5 | `HVAC_MODE_OFF` | `PRESET_NONE` |

¹ - also `MAX_DEVICE_MODE_AUTOMATIC`, but we ignore that case because schedule will override it eventually
² - any value between 5.0 and 30.0 (inclusive), stored on device when it's configured

This dictionary is just a helper and proper mapping is handled inside `preset_mode()` and `set_preset_mode()`.

MartinHjelmare replied:
Ok.

MartinHjelmare reviewed (approved):
Looks good!
